src/terminal_chat_console.cpp

```diff
@@ -0,0 +1,432 @@
+{
+	initscr();
+	cbreak(); //raw();
+	noecho();
+	keypad(stdscr, TRUE);
+	nodelay(stdscr, TRUE);
+	timeout(100);
+
+	// To make esc not delay up to one second. According to the internet,
+	// this is the value vim uses, too.
+	set_escdelay(25);
+
+	getmaxyx(stdscr, m_rows, m_cols);
+	reformat_backend(&m_chat_backend, m_rows, m_cols);
+}
+
+void TerminalChatConsole::deinit_of_curses()
+{
+	endwin();
+}
+
+void TerminalChatConsole::resize()
+{
+	m_chat_backend.reformat(m_cols, m_rows - 2);
+}
+
+void *TerminalChatConsole::run()
+{
+	BEGIN_DEBUG_EXCEPTION_HANDLER
+
+	std::cout << "========================" << std::endl;
+	std::cout << "Begin log output over terminal"
+		<< " (no stdout/stderr backlog during that)" << std::endl;
+	// Make the loggers to stdout/stderr shut up.
+	// Go over our own loggers instead.
+	LogLevelMask err_mask = g_logger.removeOutput(&stderr_output);
+	LogLevelMask out_mask = g_logger.removeOutput(&stdout_output);
+
+	g_logger.addOutput(&m_log_output);
+
+	// Inform the server of our nick
+	m_chat_interface->command_queue.push_back(
+		ChatEvent(CET_NICK_ADD, m_nick, L""));
+
+	init_of_curses();
+
+	while (!stopRequested()) {
+
+		int ch = getch();
+		if (stopRequested())
+			break;
+
+		step(ch);
+	}
+
+	deinit_of_curses();
+
+	if (m_kill_requested)
+		*m_kill_requested = true;
+
+	g_logger.removeOutput(&m_log_output);
+	g_logger.addOutputMasked(&stderr_output, err_mask);
+	g_logger.addOutputMasked(&stdout_output, out_mask);
+
+	std::cout << "End log output over terminal"
+		<< " (no stdout/stderr backlog during that)" << std::endl;
+	std::cout << "========================" << std::endl;
+
+	END_DEBUG_EXCEPTION_HANDLER
+
+	return NULL;
+}
+
+void TerminalChatConsole::typeChatMessage(const std::wstring &m)
+{
+	// Discard empty line
+	if (m == L"")
+		return;
+
+	// Send to server
+	m_chat_interface->command_queue.push_back(
+		ChatEvent(CET_CHAT, m_nick, m));
+
+	// Print if its a command (gets eaten by server otherwise)
+	if (m[0] == L'/') {
+		m_chat_backend.addMessage(L"", (std::wstring)L"Issued command: " + m);
+	}
+}
+
+void TerminalChatConsole::handleInput(int ch, bool &complete_redraw_needed)
+{
+	// Helpful if you want to collect key codes that aren't documented
+	/*if (ch != ERR) {
+		m_chat_backend.addMessage(L"",
+			(std::wstring)L"Pressed key " + utf8_to_wide(
+			std::string(keyname(ch)) + " (code " + itos(ch) + ")"));
+		complete_redraw_needed = true;
+	}//*/
+
+	switch (ch) {
+		case ERR: // no input
+			break;
+		case 27: // ESC
+			// Toggle ESC mode
+			m_esc_mode = !m_esc_mode;
+			break;
+		case KEY_PPAGE:
+			m_chat_backend.scrollPageUp();
+			complete_redraw_needed = true;
+			break;
+		case KEY_NPAGE:
+			m_chat_backend.scrollPageDown();
+			complete_redraw_needed = true;
+			break;
+		case KEY_ENTER:
+		case '\r':
+		case '\n': {
+			std::wstring text = m_chat_backend.getPrompt().submit();
+			typeChatMessage(text);
+			break;
+		} case KEY_UP:
+			m_chat_backend.getPrompt().historyPrev();
```

Comment:
Given all the m_chat_backend.getPrompt() calls, wouldn't it make sense to get the prompt only once outside of the switch statement?